### PR TITLE
Unet_2d

### PR DESCRIPTION
**Pull request by likholat**

### CODE REVIEW

File: direct/nn/unet/unet_2d.py [hunk 1]
```diff
@@ -12,6 +12,34 @@
...
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.origin_forward = super().forward
+        self.num_features = args[0]
+
+    def forward(self, input):
+        scale = torch.ones(self.num_features)
+        bias = torch.zeros(self.num_features)
```
Comment (dkurt) [inline direct/nn/unet/unet_2d.py:38]: scale and bias are unused

Comment (dkurt) [inline direct/nn/unet/unet_2d.py]: Create them inside `symbolic`

File: direct/nn/unet/unet_2d.py [hunk 2]
```diff
@@ -12,6 +12,34 @@
...
+        y = cls.origin_forward(input)
+        return y
+
+
+class InstanceNorm2dONNX(nn.InstanceNorm2d):
+    """
+    This is a support class which helps export network with InstanceNorm2d in ONNX format.
+    """
+
+    def __init__(self, *args, **kwargs):
```
Comment (dkurt) [inline direct/nn/unet/unet_2d.py]: `def __init__(self, num_features, *args, **kwargs):`

File: direct/nn/openvino/openvino_model.py [hunk 3]
```diff
@@ -11,37 +11,58 @@ def __init__(self, model):
...
-        res = self.exec_net.infer(input_map)
-        out = ([torch.Tensor(res["cell_outputs"])], torch.Tensor(res["previous_state"]))
+            net = ie.read_network(buf.getvalue(), b"", init_from_buffer=True)
+            self.exec_net = ie.load_network(net, "CPU")
 
+    def postprocess(self, res):
+        if self.model_name == "RIM":
+            out = ([torch.Tensor(res["cell_outputs"])], torch.Tensor(res["previous_state"]))
+        elif self.model_name == "Unet2d":
+            out = torch.Tensor(res["output"])
```
Comment (dkurt) [inline direct/nn/openvino/openvino_model.py:94]: Save output names into field and just iterate over them

File: tests/tests_nn/test_unet_2d.py [hunk 4]
```diff
@@ -58,4 +65,11 @@ def test_unet_2d(shape, num_filters, num_pool_layers, skip, normalized):
 
     out = model(data, sens)
 
+    if openvino_available:
+        ov_model = OpenVINOModel(model)
+        ov_out = ov_model(masked_kspace=data, sensitivity_map=sens)
```
Comment (dkurt) [inline tests/tests_nn/test_unet_2d.py]: Model launch should be equal: `ov_out = ov_model(data, sens)`

File: tests/tests_nn/test_unet_2d.py [hunk 5]
```diff
@@ -58,4 +65,12 @@ def test_unet_2d(shape, num_filters, num_pool_layers, skip, normalized):
 
     out = model(data, sens)
 
+    if openvino_available:
+        ov_model = OpenVINOModel(model)
+        # ov_out = ov_model(masked_kspace=data, sensitivity_map=sens)
+        ov_out = ov_model(masked_kspace=data)
+
+        assert torch.max(torch.abs(out[0][-1] - ov_out[0][-1])) < 1e-5
+        assert torch.max(torch.abs(out[1] - ov_out[1])) < 1e-5
```
Comment (dkurt) [inline tests/tests_nn/test_unet_2d.py]: Do we know a source of second output? According to CI, it's something related to gradients:
```
assert tensor(5.5534, grad_fn=<MaxBackward1>) < 1e-05
```
Can you try to keep only 0th check?

File: direct/nn/openvino/openvino_model.py (hunk 3 above)
Comment (likholat) [reply direct/nn/openvino/openvino_model.py:94]: Can't use this approach because outputs have different types. RIM returns `(list of Tensors, Tensor)`, Unet - just `Tensor`

File: tests/tests_nn/test_unet_2d.py (hunk 4 above)
Comment (likholat) [reply tests/tests_nn/test_unet_2d.py]: Have to do a rebase, this test has been updated

File: direct/data/transforms.py [hunk 6]
```diff
@@ -223,8 +224,10 @@ def origin_ifft2(
     assert_complex(data, complex_last=True)
 
     data = view_as_complex(data)
+
```
Comment (dkurt) [inline direct/data/transforms.py]: Avoid adding new lines in unchanged method

File: direct/data/transforms.py [hunk 7]
```diff
@@ -23,11 +23,11 @@ class FFTONNX(torch.autograd.Function):
 
     @staticmethod
     def symbolic(g, data, dim, centered, normalized, inverse=False):
-        return g.op("IFFT" if inverse else "FFT", data, centered_i=int(centered), inverse_i=int(inverse))
+        return g.op("IFFT" if inverse else "FFT", data, dim, centered_i=int(centered), inverse_i=int(inverse))
```
Comment (dkurt) [inline direct/data/transforms.py:29]:
```python
        dims = torch.tensor(dims)
        dims = g.op("Constant", value_t=dims)

        return g.op('IFFT' if inverse else 'FFT', x, dims,
                    inverse_i=inverse, centered_i=centered)
```
Using this code you don't need convert `dim` to torch.tensor in other places

File: tests/tests_nn/test_unet_2d.py [hunk 8]
```diff
@@ -58,4 +65,12 @@ def test_unet_2d(shape, num_filters, num_pool_layers, skip, normalized):
 
     out = model(data, sens)
```
Comment (dkurt) [inline tests/tests_nn/test_unet_2d.py:68]: call `model.eval()` before getting reference results

File: direct/nn/unet/unet_2d.py [hunk 9]
```diff
@@ -89,7 +120,8 @@ def __init__(self, in_channels: int, out_channels: int):
 
         self.layers = nn.Sequential(
             nn.ConvTranspose2d(in_channels, out_channels, kernel_size=2, stride=2, bias=False),
-            nn.InstanceNorm2d(out_channels),
+            # nn.InstanceNorm2d(out_channels),
```
Comment (dkurt) [inline direct/nn/unet/unet_2d.py]: remove commented lines

File: direct/nn/openvino/openvino_model.py [hunk 10]
```diff
@@ -11,37 +11,58 @@ def __init__(self, model):
...
-        }
+        if self.model_name == "RIM":
+            args = ["input_image", "masked_kspace", "sampling_mask", "sensitivity_map"]
+            output_names = ["cell_outputs", "previous_state"]
+        elif self.model_name == "Unet2d":
+            output_names = ["output"]
+            if self.model.image_initialization == "sense":
+                args = ["masked_kspace", "sensitivity_map"]
+            elif self.model.image_initialization == "zero_filled":
+                args = ["masked_kspace"]
```
Comment (dkurt) [inline direct/nn/openvino/openvino_model.py]: There are only two possible inputs in UNet2d so let's do this:
```python
args = ["masked_kspace"]
if self.model.image_initialization == "sense":
    args.append("sensitivity_map")
```
No need for exception.

File: direct/nn/openvino/openvino_model.py [hunk 11]
```diff
@@ -11,37 +11,58 @@ def __init__(self, model):
...
+                    f"Unknown image_initialization. Expected `sense` or `zero_filled`. "
+                    f"Got {self.model.image_initialization}."
+                )
 
-        if self.exec_net is None:
-            self.create_net(input_image, masked_kspace, sampling_mask, sensitivity_map)
+        if args:
+            self.input = [kwargs[k] for k in args]
+        else:
+            raise ValueError(f"The model is not supported by OpenVINO: {self.model.__class__}")
```
Comment (dkurt) [inline direct/nn/openvino/openvino_model.py]: Move this exception to `else` under `elif self.model_name == "Unet2d":`

Comment (dkurt) [inline direct/nn/openvino/openvino_model.py]: `self.model_name`

File: tests/tests_nn/test_unet_2d.py (hunk 8 above)
Comment (likholat) [reply tests/tests_nn/test_unet_2d.py:68]: Do we still need it?

**Review by dkurt — APPROVED** [review]
The error of backward grads now blames ComplexMultiplicationONNX. Let's merge this PR and rebase `ov_support` branch  first.
